core/src/main/java/org/infinispan/interceptors/compat/BaseTypeConverterInterceptor.java

```diff
@@ -268,6 +280,51 @@ public void close() {
       }
    }
 
+   @Override
+   public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
+      TypeConverter<Object, Object, Object, Object> converter = determineTypeConverter(command.getFlags());
+      CacheSet<K> set = (CacheSet<K>) super.visitKeySetCommand(ctx, command);
+
+      return new AbstractDelegatingKeyCacheSet<K, V>(getCacheWithFlags(cache, command), set) {
+
+         @Override
+         public CloseableIterator<K> iterator() {
+            return new CloseableIteratorMapper<>(super.iterator(), k -> {
+               return (K) converter.unboxKey(k);
+            });
+         }
+
+         @Override
+         public CloseableSpliterator<K> spliterator() {
+            return new IteratorAsSpliterator.Builder<>(iterator())
+                    .setEstimateRemaining(super.spliterator().estimateSize())
+                    .setCharacteristics(Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL)
+                    .get();
```

Comment:
I meant just this method, and I was thinking about implementing both `iterator()` and `spliterator()` to return the same implementation. But I guess it's not worth worrying about users mixing `next()` and `tryAdvance()`, I'm now thinking instead of implementing the iterator on top of the spliterator...